zkevm-circuits/src/evm_circuit/execution/msize.rs

```diff
@@ -0,0 +1,173 @@
+        execution::{
+            bus_mapping_tmp::{Block, Call, ExecStep, Transaction},
+            ExecutionGadget,
+        },
+        step::ExecutionState,
+        util::{
+            common_gadget::SameContextGadget,
+            constraint_builder::{
+                ConstraintBuilder, StepStateTransition, Transition::Delta,
+            },
+            from_bytes, RandomLinearCombination,
+        },
+    },
+    util::Expr,
+};
+use array_init::array_init;
+use halo2::{arithmetic::FieldExt, circuit::Region, plonk::Error};
+
+#[derive(Clone, Debug)]
+pub(crate) struct MsizeGadget<F> {
+    same_context: SameContextGadget<F>,
+    value: RandomLinearCombination<F, 8>,
+}
+
+impl<F: FieldExt> ExecutionGadget<F> for MsizeGadget<F> {
+    const NAME: &'static str = "MSIZE";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::MSIZE;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        // memory_size is limited to 64 bits so we only consider 8 bytes
+        let bytes = array_init(|_| cb.query_cell());
+        cb.require_equal(
+            "Constrain memory_size equal to stack value",
+            from_bytes::expr(&bytes),
+            cb.curr.state.memory_size.expr(),
+        );
+
+        // Push the value on the stack
+        let value = RandomLinearCombination::new(bytes, cb.randomness());
+        cb.stack_push(value.expr());
+
+        // State transition
+        let step_state_transition = StepStateTransition {
+            rw_counter: Delta(1.expr()),
+            program_counter: Delta(1.expr()),
+            stack_pointer: Delta((-1).expr()),
+            ..Default::default()
+        };
+        let opcode = cb.query_cell();
+        let same_context = SameContextGadget::construct(
+            cb,
+            opcode,
+            step_state_transition,
+            None,
+        );
+
+        Self {
+            same_context,
+            value,
+        }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        _: &Block<F>,
+        _: &Transaction<F>,
+        _: &Call<F>,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.same_context.assign_exec_step(region, offset, step)?;
+
+        self.value.assign(
+            region,
+            offset,
+            Some((step.memory_size).to_le_bytes()),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::evm_circuit::{
+        execution::bus_mapping_tmp::{
+            Block, Bytecode, Call, ExecStep, Rw, Transaction,
+        },
+        step::ExecutionState,
+        test::run_test_circuit_incomplete_fixed_table,
+        util::RandomLinearCombination,
+    };
+    use bus_mapping::{
+        eth_types::{ToBigEndian, ToLittleEndian, Word},
+        evm::OpcodeId,
+    };
+    use halo2::arithmetic::BaseExt;
+    use pairing::bn256::Fr as Fp;
+
+    #[test]
+    fn test_ok() {
+        let opcode = OpcodeId::MSIZE;
```

Comment:
Would the simplification applied here https://github.com/appliedzkp/zkevm-circuits/pull/239/files#diff-6cab7856d5161af3600f3ba7b241c0f6bb26ff41ee8d8a04595212243e76710bR93
work here as well?

Reply:
I think this PR needs rebase to work, some structures have been moved to another module.